File: app/src/main/java/com/marvelcomics/brito/marvelcomics/ui/fragment/comics/ComicsFragment.java
package com.marvelcomics.brito.marvelcomics.ui.fragment.comics;

import android.arch.lifecycle.Observer;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.marvelcomics.brito.entity.ComicEntity;
import com.marvelcomics.brito.marvelcomics.R;
import com.marvelcomics.brito.marvelcomics.databinding.FragmentComicsBinding;
import com.marvelcomics.brito.marvelcomics.ui.ResourceModelHandler;
import com.marvelcomics.brito.marvelcomics.ui.fragment.BaseFragment;
import com.marvelcomics.brito.marvelcomics.ui.fragment.ItemOffSetDecorationHorizontal;
import com.marvelcomics.brito.presentation.ResourceModel;
import com.marvelcomics.brito.presentation.viewmodel.comics.ComicsViewModel;

import java.util.List;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;

public class ComicsFragment extends BaseFragment implements ResourceModelHandler.ResourceModelListener<ResourceModel<List<ComicEntity>>> {

    private static final String ARGUMENT_CHARACTER_ID = "character_id_args";
    private int characterId;

    @Inject
    protected ComicsViewModel comicsViewModel;

    private FragmentComicsBinding binding;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        characterId = args.getInt(ARGUMENT_CHARACTER_ID);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        AndroidSupportInjection.inject(this);
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_comics, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        observeComics();
    }

    @Override
    public void onSuccessState(ResourceModel<List<ComicEntity>> resourceModel) {
        List<ComicEntity> comicResource = resourceModel.getData();
        if (!comicResource.isEmpty()) {
            createdAdapter(comicResource);
        } else {
            //TODO DialogAlert
        }
    }

    @Override
    public void onErrorState(Throwable throwable) {

    }

    @Override
    public void onLoadingState() {

    }

    private void observeComics() {
        comicsViewModel.loadComics(characterId).observe(this, new Observer<ResourceModel<List<ComicEntity>>>() {
            @Override
            public void onChanged(@Nullable ResourceModel<List<ComicEntity>> listResourceModel) {
                onModelChanged(listResourceModel, ComicsFragment.this);
            }
        });
    }

    private void createdAdapter(List<ComicEntity> comicResource) {
        ComicsAdapter adapter = new ComicsAdapter(comicResource);
        binding.recyclerviewFragmentComic.setLayoutManager(
                new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        binding.recyclerviewFragmentComic.setAdapter(adapter);
        binding.recyclerviewFragmentComic.addItemDecoration(new ItemOffSetDecorationHorizontal(8));
    }

    public static ComicsFragment newInstance(int characterId) {
        ComicsFragment fragment = new ComicsFragment();

        Bundle args = new Bundle();
        args.putInt(ARGUMENT_CHARACTER_ID, characterId);

        fragment.setArguments(args);
        return fragment;
    }
}

File: app/src/main/java/com/marvelcomics/brito/view/fragment/comics/ComicsFragment.kt
package com.marvelcomics.brito.view.fragment.comics

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.marvelcomics.brito.R
import com.marvelcomics.brito.data.handler.ResourceModel
import com.marvelcomics.brito.domain.entity.ComicEntity
import com.marvelcomics.brito.infrastructure.utils.AlertDialogUtils
import com.marvelcomics.brito.view.fragment.ItemOffSetDecorationHorizontal
import com.marvelcomics.brito.viewmodel.comic.ComicViewModel
import kotlinx.android.synthetic.main.fragment_comics.view.*
import org.koin.android.ext.android.inject

class ComicsFragment : Fragment() {

    private val comicViewModel: ComicViewModel by inject()

    private var characterId: Int? = 0

    private var progressbarLoadingComics: ProgressBar? = null
    private var recyclerviewComics: RecyclerView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        characterId = arguments?.getInt(ARGUMENT_CHARACTER_ID)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val inflatedView = inflater.inflate(R.layout.fragment_comics, null)
        progressbarLoadingComics = inflatedView.progressbar_loading_comics
        recyclerviewComics = inflatedView.recyclerview_fragment_comic
        return inflatedView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initObservers()
        loadComics()
    }

    private fun initObservers() {
        comicViewModel.comics.observe(viewLifecycleOwner, Observer {
            when (it.status) {
                ResourceModel.State.SUCCESS -> {
                    it.data?.let { listComics ->
                        showComics(listComics)
                    }
                }
                ResourceModel.State.ERROR -> {
                    it.message?.let { message ->
                        showError(message)
                    }
                }
                ResourceModel.State.LOADING -> {
                    showLoading()
                }
                else -> {

                }
            }
        })
    }

    private fun loadComics() {
        comicViewModel.characterId.value = characterId.toString()
    }

    private fun showComics(comics: List<ComicEntity>) {
        progressbarLoadingComics?.visibility = View.GONE
        recyclerviewComics?.visibility = View.VISIBLE
        createAdapter(comics)
    }

    private fun showLoading() {
        progressbarLoadingComics?.visibility = View.VISIBLE
        recyclerviewComics?.visibility = View.GONE
    }

    private fun showError(message: String) {
        progressbarLoadingComics?.visibility = View.GONE
        recyclerviewComics?.visibility = View.VISIBLE
        AlertDialogUtils.showSimpleDialog("Erro", message, requireContext())
    }

    private fun createAdapter(listComics: List<ComicEntity>) {
        val comicsAdapter = ComicsAdapter(listComics)
        recyclerviewComics?.layoutManager =
            LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        recyclerviewComics?.adapter = comicsAdapter
        recyclerviewComics?.addItemDecoration(ItemOffSetDecorationHorizontal(8))
    }

    companion object {
        const val ARGUMENT_CHARACTER_ID = "character_id_args"

        fun newInstance(characterId: Int): ComicsFragment {
            val fragment = ComicsFragment()
            val args = Bundle()
            args.putInt(ARGUMENT_CHARACTER_ID, characterId)

            fragment.arguments = args
            return fragment
        }
    }
}